### PR TITLE
Use FabricJS v1.5.0

### PR DESCRIPTION
**Pull request by pjanik**
1. https://www.pivotaltracker.com/story/show/90693952 if fixed now.

2. New FabricJS introduces some non-backward compatible changes related to paths (free hand draw tool). Resizing of paths has been updated to fix it.

3. New JSON format is incompatible with the old one (paths), so I've implement versioning of JSON state and automatic conversion from old JSON format to new one. The same pattern can be used in the future in case of need.

4. New FabricJS fixes SVG rendering, so finally we can use stamps with gradient and we don't have to provide weird tweaks to SVG images. This has a bad side effect though - if user loads old JSON with old stamps, they will look a bit incorrect. That could be theoretically handled in conversion script mentioned above, but I don't think it's worth it. Stamps will be visible, still meaningful, but some of their elements will be shifted a bit. If you try to add some stamps from "Molecules" category [on this page](http://concord-consortium.github.io/drawing-tool/examples/), you will see exactly what user is going to see if he loads his old image. "Molecules (gradient)" stamps don't have any tweaks related to old FabricJS bugs, so we should use them once this branch is merged.

This branch is deployed to:
http://concord-consortium.github.io/drawing-tool/examples/
so you can test it and take a look at stamps.

### CODE REVIEW

**Comment by ddamelin** [comment]
Not having to hand-tweak the SVG stamps is a huge improvement. It was a major pain to create the ones we are currently using, and I was discouraged from making more because of the element alignment problems.

**Comment by pjanik** [comment]
@ddamelin, by the way, when you were tweaking stamps to workaround FabricJS issues, did you find any common pattern?
If these tweaks could be generalized then I could easily workaround issue described in #4 (but I can't see any pattern at first glance).

**Comment by ddamelin** [comment]
@pjanik No. I didn't find any pattern, which is why it was so painful. Lot's of guessing and checking. Sometimes things lined up fine with no tweaks, and other times I had to futz around, hand tweaking the SVG markup. I'm sure there is a pattern, but it may relate to if the object was rotated, and the thickness of the line, and the placement of other items in the bounding box.

**Comment by ddamelin** [comment]
It certainly wasn't something as simple as always moving the lines with the same offset.

**Comment by scytacki** [comment]
@psndcsrv Can you review this? I was hoping too but I'm just too swamped.

File: app/scripts/fabric-extensions/rescale-2-resize.js [hunk 1]
```diff
@@ -40,20 +40,15 @@ var duringResize = {
...
+    var offsetY = s.pathOffset.y;
     for (var i = 0; i < s.path.length; i++) {
-      s.path[i][1] *= s.scaleX;
-      s.path[i][2] *= s.scaleY;
-      s.path[i][3] *= s.scaleX;
-      s.path[i][4] *= s.scaleY;
+      s.path[i][1] = (s.path[i][1] - offsetX) * s.scaleX + offsetX;
+      s.path[i][2] = (s.path[i][2] - offsetY) * s.scaleY + offsetY;
+      s.path[i][3] = (s.path[i][3] - offsetX) * s.scaleX + offsetX;
+      s.path[i][4] = (s.path[i][4] - offsetY) * s.scaleY + offsetY;
```
Comment (psndcsrv) [inline app/scripts/fabric-extensions/rescale-2-resize.js]: Would you need to scale offsetX/offsetY when you add it back in?
`(s.path[i][1] - offsetX) * s.scaleX + offsetX*s.scaleX`

I'm not really familiar with what's going on here, so I might be totally off.

**Comment by psndcsrv** [comment]
@scytacki @pjanik This looks fine to me. I like the chained state converter!

I made one inline comment, but I'll have to defer to your answer since I'm just looking at the code from the diffs and don't have the context about what's going on exactly.

**Comment by psndcsrv** [comment]
Whoops. They really shouldn't have a "Close and comment" button!

File: app/scripts/fabric-extensions/rescale-2-resize.js (hunk 1 above)
Comment (pjanik) [inline app/scripts/fabric-extensions/rescale-2-resize.js]: The idea is to scale path coordinates without offset. Offset remains unchanged. That's why I first subtract and then add the same value. 

You can test this behaviour if you open this page:
http://concord-consortium.github.io/drawing-tool/examples/
select free hand drawing tool (4th icon), draw anything, then select it (1st icon) and try to resize.

Comment (psndcsrv) [inline app/scripts/fabric-extensions/rescale-2-resize.js]: Ah, I see. I was thinking along the lines of scaling the entire canvas, in which case it makes sense to scale the offset, too. But this code only scales this path, so you don't need to. Got it, thanks!

**Pull request merged**